apis/r/R/write_seurat.R

```diff
@@ -0,0 +1,547 @@
+    # Write the feature loadings
+    spdl::info("Adding feature loadings as {}", sQuote(ldgs))
+    # Always write reductions as sparse arrays
+    soma_parent$varm$set(
+      object = write_soma(
+        x = mat,
+        uri = ldgs,
+        soma_parent = soma_parent$varm,
+        sparse = TRUE,
+        transpose = FALSE,
+        platform_config = platform_config,
+        tiledbsoma_ctx = tiledbsoma_ctx
+      ),
+      name = ldgs
+    )
+  }
+  return(invisible(soma_parent))
+}
+
+#' @rdname write_soma_seurat_sub
+#'
+#' @section Writing \code{\link[SeuratObject]{Graph}s}:
+#' \pkg{Seurat} \code{\link[SeuratObject]{Graph}} objects are
+#' written out as \link[tiledbsoma:SOMASparseNDArray]{sparse matrices}
+#' to the \dQuote{\code{obsp}} group of a
+#' \link[tiledbsoma:SOMAMeasurement]{measurement}
+#'
+#' @method write_soma Graph
+#' @export
+#'
+write_soma.Graph <- function(
+  x,
+  uri,
+  soma_parent,
+  ...,
+  platform_config = NULL,
+  tiledbsoma_ctx = NULL,
+  absolute = FALSE
+) {
+  .check_seurat_installed()
+  stopifnot(
+    "'soma_parent' must be a SOMAMeasurement" = inherits(
+      x = soma_parent,
+      what = 'SOMAMeasurement'
+    ),
+    "'absolute' must be a single logical value" = is_scalar_logical(absolute)
+  )
+  if (!'obsp' %in% soma_parent$names()) {
+    soma_parent$obsp <- SOMACollectionCreate(
+      uri = file_path(soma_parent$uri, 'obsp'),
+      platform_config = platform_config,
+      tiledbsoma_ctx = tiledbsoma_ctx
+    )
+  }
+  soma_parent$obsp$set(
+    object = NextMethod(
+      generic = 'write_soma',
+      object = x,
+      uri = uri,
+      soma_parent = soma_parent$obsp,
+      sparse = TRUE,
+      transpose = FALSE,
+      platform_config = platform_config,
+      tiledbsoma_ctx = tiledbsoma_ctx
+    ),
+    name = uri
+  )
+  return(invisible(soma_parent))
+}
+
+#' Write a \code{\link[SeuratObject]{Seurat}} object to a SOMA
+#'
+#' @inheritParams write_soma
+#' @param x A \code{\link[SeuratObject]{Seurat}} object
+#' @param overwrite Overwrite existing data located at \code{uri}
+#'
+#' @inherit write_soma return
+#'
+#' @section Writing Cell-Level Meta Data:
+#' Cell-level meta data is written out as a
+#' \link[tiledbsoma:SOMADataFrame]{data frame} called \dQuote{\code{obs}} at
+#' the \code{\link[tiledbsoma:SOMAExperiment]{experiment}} level
+#'
+#' @inherit write_soma_seurat_sub_objects sections
+#'
+#' @method write_soma Seurat
+#' @export
+#'
+write_soma.Seurat <- function(
+  x,
+  uri = NULL,
+  ...,
+  platform_config = NULL,
+  tiledbsoma_ctx = NULL,
+  overwrite = FALSE
+) {
+  .check_seurat_installed()
+  stopifnot(
+    "'uri' must be a single character value" = is.null(uri) ||
+      is_scalar_character(uri)
+  )
+  uri <- uri %||% file_path(user_dir(), SeuratObject::Project(x))
+  if (!is_remote_uri(uri)) {
+    if (isTRUE(overwrite)) {
+      unlink(x = uri, recursive = TRUE, force = TRUE)
+    }
+    dir.create(dirname(uri), recursive = TRUE)
+  }
```

Comment:
This will silently skip the deletion step with remote URIs. I don't think we want to implement anything that results in different behavior for local vs remote locations. Adding an overwrite option is also something we should discuss to make sure we're all on the same page and we implement it in a consistent way across the codebase.

Reply:
Correction:

> have differences _only when there are language-specific or toolchain-specific needs to do so_